packages/gatsby-source-filesystem/src/file-watcher.js

```diff
@@ -0,0 +1,208 @@
+const chokidar = require(`chokidar`)
+const { Machine } = require(`xstate`)
+
+const { createFileNode } = require(`./create-file-node`)
+/**
+ * Create a state machine to manage Chokidar's not-ready/ready states and for
+ * emitting file system events into Gatsby.
+ *
+ * On the latter, this solves the problem where if you call createNode for the
+ * same File node in quick succession, this can leave Gatsby's internal state
+ * in disarray causing queries to fail. The latter state machine tracks when
+ * Gatsby is "processing" a node update or when it's "idle". If updates come in
+ * while Gatsby is processing, we queue them until the system returns to an
+ * "idle" state.
+ */
+const createFSMachine = () =>
+  Machine({
+    key: `emitFSEvents`,
+    parallel: true,
+    strict: true,
+    states: {
+      CHOKIDAR: {
+        initial: `CHOKIDAR_NOT_READY`,
+        states: {
+          CHOKIDAR_NOT_READY: {
+            on: {
+              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
+              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+            },
+          },
+          CHOKIDAR_WATCHING: {
+            on: {
+              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
+            },
+          },
+          CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED: {
+            on: {
+              CHOKIDAR_READY: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+            },
+          },
+        },
+      },
+      PROCESSING: {
+        initial: `BOOTSTRAPPING`,
+        states: {
+          BOOTSTRAPPING: {
+            on: {
+              BOOTSTRAP_FINISHED: `IDLE`,
+            },
+          },
+          IDLE: {
+            on: {
+              EMIT_FS_EVENT: `PROCESSING`,
+            },
+          },
+          PROCESSING: {
+            on: {
+              QUERY_QUEUE_DRAINED: `IDLE`,
+              TOUCH_NODE: `IDLE`,
+            },
+          },
+        },
+      },
+    },
+  })
+
+const fsMachine = createFSMachine()
+let currentState = fsMachine.initialState
```

Comment:
above 2 lines ashould be inside`fileWatcher` function - otherwise there would be single instance of state machine even if multiple instances of plugin are running which would cause problems